xcdat/dataset.py

```diff
@@ -203,46 +195,25 @@ def open_mfdataset(
 
     .. [2] https://xarray.pydata.org/en/stable/generated/xarray.open_mfdataset.html
     """
-    # `xr.open_mfdataset()` drops the time coordinates encoding dictionary if
-    # multiple files are merged with `decode_times=True` (refer to
-    # https://github.com/pydata/xarray/issues/2436). The workaround is to store
-    # the time encoding from the first dataset as a variable, and add the time
-    # encoding back to final merged dataset.
-    time_encoding = None
-
-    if decode_times:
-        time_encoding = _keep_time_encoding(paths)
-
-        cf_compliant_time: Optional[bool] = _has_cf_compliant_time(paths)
-        # xCDAT attempts to decode non-CF compliant time coordinates using the
-        # preprocess keyword arg with `xr.open_mfdataset()`.
-        if cf_compliant_time is False:
-            decode_times = False
-            preprocess = partial(_preprocess_non_cf_dataset, callable=preprocess)
-
     ds = xr.open_mfdataset(
         paths,
-        decode_times=decode_times,
+        decode_times=False,
         data_vars=data_vars,
-        preprocess=preprocess,
         **kwargs,  # type: ignore
     )
-    ds = _postprocess_dataset(ds, data_var, center_times, add_bounds, lon_orient)
 
-    if time_encoding is not None:
-        time_dim = get_axis_dim(ds, "T")
-        ds[time_dim].encoding = time_encoding
-        # Update "original_shape" to reflect the final time coordinates shape.
-        ds[time_dim].encoding["original_shape"] = ds[time_dim].shape
+    ds = _postprocess_dataset(
+        ds, data_var, decode_times, center_times, add_bounds, lon_orient
+    )
 
     return ds
```

Comment:
@pochedley@gmail.com Refactoring `.decode_non_cf_time()`to `decode_time()` turns out to be a big win.

Using `decode_time()` to decode both CF and non-CF units as `cftime` objects gives us these benefits:

1. Time encoding info is no longer lost with `open_mfdataset()`, since `decode_times=False` preserves them.
2. No longer need to check CF-compliance in the first file, which requires significant logic.
3. Instead of decoding time coordinates one file at a time, we can merge all files into a single Dataset THEN decode time coordinates in a single operation. This is a potential performance gain.
4. Eliminate helper functions like `_keep_time_encoding()`, `_has_cf_compliant_time()`, etc. for cleaner code.
5. More predictability, since time coordinates will always be `cftime` if decoded with `xcdat`.

Reply:
@tomvothecoder - regarding point 3 – one issue I've hit a couple times is that xarray seems to not open mfdatasets with mixed time units:

```
import xcdat as xc
dpath = '/p/css03/esgf_publish/CMIP6/PMIP/MIROC/MIROC-ES2L/past1000/r1i1p1f2/Amon/tas/gn/v20200318/*nc'
ds = xc.open_mfdataset(dpath, use_cftime=True)
```

> TypeError: Cannot combine along dimension 'time' with mixed types. Found: Datetime360Day, DatetimeProlepticGregorian. If importing data directly from a file then setting `use_cftime=True` may fix this issue.

I'm wondering if our preprocessing operations might address this by individually decoding netCDF files using cftime and then merging them? 

This issue affects `xarray` and may be complicated to address, but I raise it because if we did prioritize it, it would likely be via preprocessing (so we may not want to keep some of that architecture).

Reply:
Thanks for pointing the possibility of datasets using different calendars (which I think is not correct or good practice?).

I ran some experiments to try to open up those files in native `xarray` with different settings.

## 1. Keep time coordinates as integers (`decode_times=False`)

* Code: 
	```python
	import xarray as xr
	
	dpath = '/p/css03/esgf_publish/CMIP6/PMIP/MIROC/MIROC-ES2L/past1000/r1i1p1f2/Amon/tas/gn/v20200318/*nc'
	
	ds= xr.open_mfdataset(dpath, decode_times=False)
	```

* Produces error: `ValueError: Resulting object does not have monotonic global indexes along dimension time`
* Why is this happening?
	>This isn't a bug, it's throwing the error it should be throwing given your input. However I can see how the documentation doesn't make it very clear as to why this is happening!
	>...
    >
	> That error is an explicit rejection of the input you gave it: "you gave me overlapping coordinates, I don't know how to concatenate those, so I'll reject them." Normally this makes sense - when the overlapping coordinates aren't NaNs then it's ambiguous as to which values to choose.
    > ...
    >
    > Instead, what you need to do is trim off the overlap first. That shouldn't be hard - presumably you know (or can determine) how big your overlap is, and all your NaNs are on one dataset. You just need to use the .isel() method with a slice. Once you've got rid of the overlapping NaNs then you should be able to combine it fine (and you shouldn't need to specify compat either). If you're using combine_by_coords as part of opening many files with open_mfdataset then it might be easier to write a trimming function which you apply first using the preprocess argument to open_mfdataset.
    >  --  https://stackoverflow.com/a/62860066

**Based on the above, it seems like handling this edge case using `preprocess` might be beyond the scope of `xcdat` (data quality, possibly complex workaround).**  

## 2. Try to decode time coordinates as datetimes (`decode_times=True`) 

* Code: 
	```python
	import xarray as xr
	
	dpath = '/p/css03/esgf_publish/CMIP6/PMIP/MIROC/MIROC-ES2L/past1000/r1i1p1f2/Amon/tas/gn/v20200318/*nc'
	
	ds = xr.open_mfdataset(dpath, decode_times=True)
	```
* Produces warning
	```python
	    /home/vo13/miniconda3/envs/xcdat_dev/lib/python3.9/site-packages/xarray/coding/times.py:150: SerializationWarning: Ambiguous reference date string: 850-1-1. The first value is assumed to be the year hence will be padded with zeros to remove the ambiguity (the padded reference date string is: 0850-1-1). To remove this message, remove the ambiguity by padding your reference date strings with zeros.
	    warnings.warn(warning_msg, SerializationWarning)
	    /home/vo13/miniconda3/envs/xcdat_dev/lib/python3.9/site-packages/xarray/coding/times.py:682: SerializationWarning: Unable to decode time axis into full numpy.datetime64 objects, continuing using cftime.datetime objects instead, reason: dates out of range
	    dtype = _decode_cf_datetime_dtype(data, units, calendar, self.use_cftime)
	    /home/vo13/miniconda3/envs/xcdat_dev/lib/python3.9/site-packages/xarray/core/indexing.py:524: SerializationWarning: Unable to decode time axis into full numpy.datetime64 objects, continuing using cftime.datetime objects instead, reason: dates out of range
	    return np.asarray(array[self.key], dtype=None)
	```

* Produces error: `TypeError: Cannot combine along dimension 'time' with mixed types. Found: Datetime360Day, DatetimeProlepticGregorian. If importing data directly from a file then setting `use_cftime=True` may fix this issue.`

### # 3. Try to decode time coordinates as datetimes (decode_times=True)

* Code:
	```python
	import xarray as xr
	
	dpath = '/p/css03/esgf_publish/CMIP6/PMIP/MIROC/MIROC-ES2L/past1000/r1i1p1f2/Amon/tas/gn/v20200318/*nc'
	
	ds = xr.open_mfdataset(dpath, decode_times=True, use_cftime=True)
	```

* Produces error: `TypeError: Cannot combine along dimension 'time' with mixed types. Found: Datetime360Day, DatetimeProlepticGregorian. If importing data directly from a file then setting 'use_cftime=True' may fix this issue.`

Reply:
@tomvothecoder – thanks for looking into this (that actually wasn't my intention). I don't think we should solve this issue now – I mostly just wanted to say that we should _try to_ leave open the possibility of addressing this kind of issue in the future. I didn't realize that time steps actually overlap (I thought it was just calendar differences). Ideally we _could_ handle calendar differences (this would be future work when it becomes enough of a priority for someone to try to solve this with a PR).

Reply:
No worries Steve, I thought it made sense to take a look into it since the code I'm writing is directly related. It would be nice to be able to address it in this PR (if it made sense).

Sounds good, let's revisit in the future.

Reply:
Moved this discussion to https://github.com/xCDAT/xcdat/discussions/361